src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs

```diff
@@ -62,7 +82,37 @@ protected override async Task RemoveKeyAsync(string cacheKey)
         /// (account or app).</returns>
         protected override async Task<byte[]> ReadCacheBytesAsync(string cacheKey)
         {
-            return await _distributedCache.GetAsync(cacheKey).ConfigureAwait(false);
+            var startTicks = Utility.Watch.Elapsed.Ticks;
+
+            // check memory cache first
+            byte[] result = (byte[])_memoryCache.Get(cacheKey);
+            _logger.LogDebug($"[IdWebCache] MemoryCache: Read cache {cacheKey} Byte size: {result?.Length}. ");
+
+            if (result == null)
+            {
+                // not found in memory, check distributed cache
+                result = await _distributedCache.GetAsync(cacheKey).ConfigureAwait(false);
+                _logger.LogDebug($"[IdWebCache] DistributedCache read: No result in memory, distributed cache result - Byte size: {result?.Length}. ");
+
+                // back propagate to memory cache
+                if (result != null)
+                {
+                    MemoryCacheEntryOptions memoryCacheEntryOptions = new MemoryCacheEntryOptions()
+                    {
+                        AbsoluteExpirationRelativeToNow = _expirationTime,
+                        Size = result?.Length,
+                    };
+
+                    _logger.LogDebug($"[IdWebCache] Back propagate from Distributed to Memory, Byte size: {result?.Length}");
+                    _memoryCache.Set(cacheKey, result, memoryCacheEntryOptions);
+                    _logger.LogDebug($"[IdWebCache] MemoryCache: Count: {_memoryCache.Count}");
+                }
+            }
+
+            _logger.LogDebug($"[IdWebCache] Read caches for {cacheKey} returned Bytes {result?.Length} Time in Ticks: {Utility.Watch.Elapsed.Ticks - startTicks}. ");
+#pragma warning disable CS8603 // Possible null reference return.
```

Comment:
You could return an empty byte[] (i.e. new byte[0]) instead of null and get rid of this warning.

Reply:
What's the contract with MSAL.NET, @bgavrilMS : I thought that null needed to be returned when the cache does not exist?

here this is because Ms.Id.Web overrides an MSAL.NET method that does not use nullable (and therefore the signature does no have it)

Reply:
my understanding was to return null as well, at least that's the current behavior.

---
In reply to: [586403545](https://github.com/AzureAD/microsoft-identity-web/pull/1037#discussion_r586403545) [](ancestors = 586403545)